python/example_code/rds/lending_library/library_api/chalicelib/library_data.py

```diff
@@ -0,0 +1,397 @@
+    def from_env(cls):
+        """
+        Creates a storage object based on environment variables.
+        """
+        cluster_name = os.environ.get('CLUSTER_NAME', '')
+        secret_name = os.environ.get('SECRET_NAME', '')
+        db_name = os.environ.get('DATABASE_NAME', '')
+        cluster = boto3.client('rds').describe_db_clusters(
+            DBClusterIdentifier=cluster_name)['DBClusters'][0]
+        secret = boto3.client('secretsmanager').describe_secret(
+            SecretId=secret_name)
+        rdsdata_client = boto3.client('rds-data')
+        return cls(cluster, secret, db_name, rdsdata_client)
+
+    def _begin_transaction(self):
+        """
+        Begins a database transaction.
+
+        :return: The transaction ID.
+        """
+        result = self._rdsdata_client.begin_transaction(
+            database=self._db_name,
+            resourceArn=self._cluster['DBClusterArn'],
+            secretArn=self._secret['ARN'])
+        return result['transactionId']
+
+    def _commit_transaction(self, transaction_id):
+        """
+        Commits a database transaction.
+
+        :return: The result of committing the transaction.
+        """
+        result = self._rdsdata_client.commit_transaction(
+            resourceArn=self._cluster['DBClusterArn'],
+            secretArn=self._secret['ARN'],
+            transactionId=transaction_id)
+        return result['transactionStatus']
+
+    def _rollback_transaction(self, transaction_id):
+        """
+        Rolls back a database transaction.
+
+        :return: The result of rolling back the transaction.
+        """
+        result = self._rdsdata_client.rollback_transaction(
+            resourceArn=self._cluster['DBClusterArn'],
+            secretArn=self._secret['ARN'],
+            transactionId=transaction_id)
+        return result['transactionStatus']
+
+    def _run_statement(self, sql, sql_params=None, transaction_id=None):
+        """
+        Runs a SQL statement and associated parameters using RDS Data Service.
+
+        :param sql: The SQL statement to run.
+        :param sql_params: The parameters associated with the SQL statement.
+        :transaction_id: The ID of a previously created transaction.
+        :return: The result of running the SQL statement.
+        """
+        try:
+            run_args = {
+                'database': self._db_name,
+                'resourceArn': self._cluster['DBClusterArn'],
+                'secretArn': self._secret['ARN'],
+                'sql': sql
+            }
+            if sql_params is not None:
+                run_args['parameters'] = sql_params
+            if transaction_id is not None:
+                run_args['transactionId'] = transaction_id
+            result = self._rdsdata_client.execute_statement(**run_args)
+            logger.info("Ran statement on %s.", self._db_name)
+        except ClientError as error:
+            if error.response['Error']['Code'] == 'BadRequestException' and \
```

Comment:
This could also be wrapped in parens for better formatting.

Reply:
Fixed.